### ISSUE TITLE
ADT-style enums/user-defined types

### ISSUE THREAD

**Issue by stephenh** [mikro-orm/mikro-orm#274]
I'd like to model enums as ADTs, i.e. something like:

```typescript
export type ProjectState = 
  | { kind: "active", id: 1, name: "Active", maybeSomething: "else" }
  | { kind: "draft", id: 2, name: "Draft", ... };

class Project {
  @EnumButAnAdt
  state: ProjectState
}
```

I know (assume) this isn't support now; do you have suggestions on how I might be able to add it? Looks easy, looks hard?

I suppose this is a specific instance of a user-defined type, like if there was an `@UserType`-type function where I provided "how to turn db into JS value" and "how to turn JS value into db", i.e.:

```
@UserType({ fromDb: int => lookupAdtValue(int); toDb: adtValue => adtValue.id })
state: ProjectState
```

**Comment by B4nan** [comment]
And the definition is in database or you have that locally in the code? If its in the database, why not use `@ManyToOne`? If its local, you could have define the property as `get/set` and handle the transformation there, or have it as numeric enum and add virtual property that would change the enum value, but act as an object:

https://mikro-orm.io/docs/defining-entities/#virtual-properties

Could you provide more information about how it is supposed to be used? 

Btw the `@UserType` thing is also on my roadmap (value transformers), but should be possible to achieve that via `get/set` right now I think.

**Comment by B4nan** [comment]
I am planning to support something like this, that should be fine for you, right? 

https://www.doctrine-project.org/projects/doctrine-orm/en/latest/cookbook/custom-mapping-types.html

**Comment by B4nan** [comment]
Here is initial implementation for that, will try to finish that soon: #276.

**Comment by B4nan** [comment]
Custom types are now available in `3.0.0-rc.5`: https://mikro-orm.io/docs/custom-types/

**Comment by stephenh** [comment]
@B4nan sorry for going silent on this; the custom types looks great! I'll poke around at those in rc.5 and follow up if I ran into anything weird about them, but so far from the docs they look very straightforward.
